designs/webviews/webviews-general.md

```diff
@@ -0,0 +1,95 @@
+
+## Current State
+
+VS Code offers a very limited set of UI functionality. Specifically, VS Code offers the following constructs for user-interactible workflows:
+
+* [Quick Picks](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#quick-pick), which present a list of options for a user to select. Users can select one or multiple items and can filter the list but cannot add to it.
+* [Input Boxes](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#quick-pick), which allow a user to submit a string.
+* [File Pickers](https://code.visualstudio.com/api/extension-capabilities/common-capabilities#file-picker), which allow a user to choose a file or directory.
+* [Modal/Non-Modal/Status Bar Messages](https://code.visualstudio.com/api/references/vscode-api#window), which can contain arbitrary buttons.
+* [Tree Views](https://code.visualstudio.com/api/references/vscode-api#TreeView), representing hierarchical information.
+
+The workflow-centric constructs (quick picks, input boxes, and file pickers) can only be displayed one at a time, which can lead to confusion in workflows where users are sequentially asked for inputs. For instance, a previous step's selection is invalid could cause a current selection to become invalid, Backtracking through the workflow to correct the mistake (or simply change values) is a pain point. These workflows also remove the user's context, forcing them to backtrack to remember what they entered previously.
+
+Another major reported pain point from working within the VS Code-specific constructs come from some of the confusion with how the inputs look the same but register differently depending on their type; an example can be seen in [the following Github issue](https://github.com/aws/aws-toolkit-vscode/issues/650).
+
+Our toolkit does currently include some webviews based on Vue.js. These webviews for the most part get the job done but have an inconsistent design and have little to no styling. These webviews are also written in Javascript, eschewing any type safety. Furthermore, they use a relatively esoteric Lodash-based templating system to create their HTML. We have limited their use within the toolkit, both due to these facts as well as a desire to stay as close to native VS Code constructs as possible; the only workflows that use these are for remote AWS Lambda invocations and EventBridge Schema searching, which both feature use-cases that are too complex for the workflow constructs.
+
+All of our other IDE toolkits feature interactibility via form-like interfaces, showing multiple fields at a time and allowing for validation of individual elements upon submission without having to backtrack through multiple steps.
+
+## Proposed Solution
+
+I propose we move more complex UIs to VS Code webviews, with "more complex" defined as:
```

Comment:
For a generic design document that's potentially long lived, don't use 'first person'.

Reply:
Also, the PR is the proposal. The document should read as factual.